src/main/java/de/retest/recheck/TestCaseFinder.java

```diff
@@ -32,51 +37,128 @@ private TestCaseFinder() {}
+
+	private static Optional<String> findTestCaseMethodInStack( final Function<TestCaseInformation, String> mapper ) {
+		final TestCaseInformation info = findTestCaseMethodInStack();
+		return info.isFound() ? Optional.of( mapper.apply( info ) ) : Optional.empty();
+	}
+
+	private static Optional<String> findTestCaseMethodInStack( final Function<TestCaseInformation, String> mapper,
+			final StackTraceElement[] trace ) {
+		final TestCaseInformation info = findTestCaseMethodInStack( trace );
+		return info.isFound() ? Optional.of( mapper.apply( info ) ) : Optional.empty();
+	}
+
+	/**
+	 * @return Test case information for the test case method in all stack traces.
+	 */
+	public static TestCaseInformation findTestCaseMethodInStack() {
 		for ( final StackTraceElement[] stack : Thread.getAllStackTraces().values() ) {
-			final StackTraceElement testCaseStackElement = findTestCaseMethodInStack( stack );
-			if ( testCaseStackElement != null ) {
-				return testCaseStackElement;
+			final TestCaseInformation info = findTestCaseMethodInStack( stack );
+			if ( info.isFound() ) {
+				return info;
 			}
 		}
-		return null;
+		return NO_TEST_CASE_INFORMATION;
 	}
 
-	public static StackTraceElement findTestCaseMethodInStack( final StackTraceElement[] trace ) {
-		boolean inTestCase = false;
-		for ( int i = 0; i < trace.length; i++ ) {
-			if ( isTestCase( trace[i] ) ) {
-				inTestCase = true;
-			} else if ( inTestCase ) {
-				return trace[i - 1];
+	/**
+	 * @param trace
+	 *            The trace to be used for search.
+	 * @return Test case information for the test case method in the given stack trace.
+	 */
+	public static TestCaseInformation findTestCaseMethodInStack( final StackTraceElement[] trace ) {
+		for ( final StackTraceElement element : trace ) {
+			final TestCaseAnnotationType type = determineTestCaseAnnotationType( element );
+			if ( type == TestCaseAnnotationType.NORMAL ) {
+				return new TestCaseInformation( element, type, 1 );
+			}
+			if ( type == TestCaseAnnotationType.REPEATABLE ) {
+				final int count = repeatableTestCaseAnnotationsCount.merge( element, 1, Math::addExact );
+				return new TestCaseInformation( element, type, count );
 			}
 		}
-		return null;
+		return NO_TEST_CASE_INFORMATION;
 	}
 
-	private static boolean isTestCase( final StackTraceElement element ) {
+	private static TestCaseAnnotationType determineTestCaseAnnotationType( final StackTraceElement element ) {
 		final Method method = tryToFindMethodForStackTraceElement( element );
 		if ( method == null ) {
-			return false;
+			return TestCaseAnnotationType.NONE;
 		}
 
 		final Annotation[] annotations = method.getAnnotations();
 		for ( final Annotation annotation : annotations ) {
 			final String annotationName = annotation.annotationType().getName();
 			if ( testCaseAnnotations.contains( annotationName ) ) {
-				return true;
+				return TestCaseAnnotationType.NORMAL;
+			}
+			if ( repeatableTestCaseAnnotations.contains( annotationName ) ) {
```

Comment:
Change order.

Reply:
I think I'll keep it as it is, natural order. Having two annotations is just wrong usage I guess.